feat(ui): tabbed canvases

csehatt741 opened the pull request:
## Summary

A new feature was implemented to enable multiple canvases in the canvas editor.

- `SliceConfig` was updated in `types.js` to handle partially undoable slices:
  - new generic parametes, `TInternalState` and `TSerializedState`, were added to strongly type states used in Redux and the persisted into storage
  - new functions, `wrapState` and `unwrapState`, were added to the `migrate` field for wrapping/unwrapping state during serialization/deserialization
  - `undoableConfig` was deleted, as `reduxUndoOptions` is used only in slices, so this field became redundant
- `store.ts` was refactored to use the updated `SliceConfig`
- new schemas, `zStateWithHistory`, `zCanvasStateWithHistory`, `zCanvasesState`, `zCanvasesStateWithHistory` and `zCanvasesStateWithoutHistory`, were created in `types.ts` to represent types in the partially undoable canvases slice
- new selectors, `selectCanvases` `selectSelectedCanvasWithHistory` and `selectSelectedCanvas`, were added to `selectors.ts` to isolate changes due to refactoring in the canvas slice from components
- `canvasSlice` was split into two parts representing the `canvases` slice without history and the `canvas` with undoable history
- `undoableCanvasesReducer` higher order reducer was created to combine `canvases` and `canvas` reducers
- `RTK` was updated to version `2.9.0`

## Related Issues / Discussions

Closes #8380

## Checklist

- [x] _The PR has a short but descriptive title, suitable for a changelog_
- [ ] _Tests added / updated (if applicable)_
- [ ] _❗Changes to a redux slice have a corresponding migration_
- [ ] _Documentation added / updated (if applicable)_
- [ ] _Updated `What's New` copy (if doing a release after this PR)_

csehatt741 commented:
@psychedelicious, here is the current status of multi-canvas development and refactoring:
- to limit the scope of refactoring I still kept the notion of the active canvas
  - it always provides scope for any components even not in `CanvasInstanceContext`, without refactoring the component hierarchy
- `canvasSlice`:
  - state is split into per canvas instance state
  - undo/redo per canvas implemented
- `canvasStagingAreaSlice` state is split into per canvas instance state
- `canvasSettingsSlice` state is split into shared and per canvas instance state

csehatt741 commented:
Since there is always only one canvas rendered, the active one, so the `CanvasManager` is created and initialized when the canvas is displayed and destroyed when the canvas becomes inactive.
Though, I haven't seen any issues caused by this lifecycle of the `CanvasManager` yet, but I expect this might lead to problems:
 - what is the effect of destroying a `CanvasManager` when the canvas becomes inactive and then reusing it when the canvas goes active again: https://github.com/invoke-ai/InvokeAI/pull/8553/files#diff-59388fa7d596ff6926d86545074a35049822341021aa0fb6df10e5a6f512ce12R48
 - how about queued and pending tasks when the UI is reopened? Do they expect the `CanvasManager` is available when they complete?

psychedelicious commented:
After more contemplation and some experimentation, I think consolidating the slices into canvas instances is the right move. That said, I'm certainly open to discussion and being wrong about this!

Here's a draft proposal for consolidated canvas state. I think it covers most of the required touch points, except interacting with the `CanvasManager` class.

Consolidated root canvas state and instance state structure:
```ts
type RootCanvasState = {
  _version: number;
  activeCanvasId: string;
  canvases: Record<string, CanvasInstance>;
};

type CanvasInstance = {
  id: string;
  name: string;
  canvas: StateWithHistory<CanvasEntityState>; // previously `CanvasState`
  params: ParamsState;
  settings: SettingsState;
  staging: StagingState;
};
```

Utilities for working with it:
```ts
const isCanvasInstanceAction = isAnyOf(...Object.values(canvasSlice.actions).map((a) => a.match));

// use a Symbol to prevent key collisions
const CANVAS_ID_META_KEY = Symbol('canvasId');

const injectCanvasId = (canvasId: string, action: CanvasInstanceAction) => {
  // mutates the action - could inject into payload instead
  Object.assign(action, { meta: { [CANVAS_ID_META_KEY]: canvasId } });
}

const extractCanvasId = (action: CanvasInstanceAction): string | undefined => {
  return action?.meta?.[CANVAS_ID_META_KEY];
}
```

Use react context to provide current canvasId:
```ts
const CanvasIdContext = createContext<string | null>(null);
const useCanvasIdContextSafe = () => useContext(CanvasIdContext);
const useCanvasIdContextRequired = () => {
  const canvasId = useCanvasIdContext();
  if (canvasId === null) {
    throw new Error("useCanvasIdContextRequired must be used within a CanvasIdProvider");
  }
  return canvasId;
}
```

Wrapper around useDispatch to auto-inject canvasId (could also be middleware I suppose):
```ts
const useAppDispatch = () => {
  const _dispatch = useDispatch();
  const canvasId = useCanvasIdContextSafe() // string | null
  return useCallback((action: UnknownAction) => {
    if (canvasId !== null && isCanvasInstanceAction(action)) {
      injectCanvasId(canvasId, action)
    }

    dispatch(action)
  })
}
```

Wrap the undoable slice reducers:
```ts
const canvasEntitySlice = createSlice(...)
const undoableConfig = { /* as before */ }
export const undoableCanvasEntityReducer = undoable(canvasEntitySlice.reducer, undoableConfig)
```

Use an "action router" pattern in the root canvas reducer to route actions to the correct canvas instance:
```ts
extraReducers: (builder) => {

  builder.addMatcher(isCanvasInstanceAction, (state, action) => {
    const canvasId = extractCanvasId(action);
    if (!canvasId) {
      return;
    }

    const instance = state.canvases[canvasId];
    if (!instance) {
      return;
    }

    // delegate to sub-reducers (each handles own undo/redo)
    if (isCanvasEntityAction(action)) {
      instance.canvas = undoableCanvasEntityReducer(instance.canvas, action);
    }
    if (isParamsAction(action)) {
      instance.params = paramsReducer(instance.params, action);
    }
    // etc
  });
}
```

To handle key-specific persist denylists, we could update `PersistConfig` to be something like this, instead of having a denylist and methods to wrap history:
```ts
type PersistConfig<TInternal, TSerialized> = {
  serialize?: (state: TInternal) => TSerialized;
  deserialize?: (state: TSerialized) => TInternal;
}

const wrapHistory = <T,>(state: T): StateWithHistory<T> => ({ /* ... */ })
const unwrapHistory = <T,>(state: StateWithHistory<T>): T => ({ /* ... */ })

const canvasEntityPersistConfig: PersistConfig<StateWithHistory<CanvasEntityState>, CanvasEntityState> = {
  migrate: (state) => zConsolidatedCanvasState.parse(state),

  serialize: (state) => ({
    _version: state._version,
    activeCanvasId: state.activeCanvasId,
    canvases: Object.fromEntries(
      Object.entries(state.canvases).map(([id, instance]) => [
        id,
        {
          ...instance,
          canvas: unwrapHistory(instance.canvas)
          params: omit(instance.params, ['sensitiveKey']) // key-specific denylist
          staging: undefined, // don't persist staging (can't remember if this is needed in the actual codebase but you get the idea)
        }
      ])
    ),
  }),

  deserialize: (state, initialState) => ({
    _version: state._version,
    activeCanvasId: state.activeCanvasId,
    canvases: Object.fromEntries(
      Object.entries(state.canvases).map(([id, instance]) => [
        id,
        {
          ...instance,
          canvas: wrapHistory(instance.canvas),
          params: merge(instance.params, defaults), // inject defaults
          staging: getInitialStagingState(id), // re-initialize staging
        }
      ])
    ),
  }),
}
```

In components, to select state, so long as we use `createSelector`, we'll get weakmap-memoized selector caches for free:
```ts
export const selectCanvasInstance = createSelector(
  [
    (state: RootState) => state.canvas.canvases,
    (state: RootState, canvasId: string) => canvasId
  ],
  (canvases, canvasId) => canvases[canvasId]
);

export const selectCanvasSettings = createSelector(
  [selectCanvasInstance],
  (instance) => instance.settings
);

// because this is composed w/ the other selectors, it will have a second arg for canvasId
export const selectBrushWidth = createSelector(
  [selectCanvasSettings],
  (settings) => settings.brushWidth
);

// in the component
const canvasId = useCanvasIdContextRequired();
const brushWidth = useAppSelector((state) => selectBrushWidth(state, canvasId));
```

csehatt741 commented:
@psychedelicious
Actually I've explicitly invested into migration, but even did not try it out with a clean install before.
The root cause was missing canvas states in auxiliary canvas slices when the canvas is initialized the first time.
Issue has been fixed and PR updated.

1. I just also had the same idea, as I was making progress with this feature and had to cope with multiple related issues, so wanted to discuss merging related slices into `canvasSlice`.
The issue above was definitely related to this problem too.

The reason why I have not combined these slices yet are, because I want to reduce the amount of code change and I just realized the complexity of issues caused by keeping the current design yesterday. I did not have time to think it over and discuss earlier.

I'll investigate this issue and will get back soon!

2. I chose not to implicitly encapsulate context information in `useAppDispatch` to give the developer the option to explicitly deal with other contexts different from the active one, eg.: to dispatch actions to any canvases. Maybe it's unnecessary, as there's always only one canvas rendered.

3. I've tried out invoking the canvas instance reducer in `extraReducers` as suggested, but undo/redo did not work there. It might be related to proxies within reducers. So I explicitly created the `undoableCanvasesReducer` invoking `canvasesSlice` and `undoableCanvasReducer` there.

csehatt741 commented:
@psychedelicious,
I've made the following progress based on your guidance:
- refactored and merged slices:
  - merged canvas, params, settings, staging into `canvasSlice`
  - created `tabSlice` and merged loras, params and refImages into it
- created `actionContextMiddleware`, which is responsible for injecting context into the action. Context is extracted by canvas and tab slices and used for action routing
- refactored `PersistConfig`

To reduce the scope of this PR I removed `CanvasInstanceContext` and made selectors used by controls based only on the active canvas. I believe this simplified multi-canvas architecture should cover most of the use-cases, which could be improved in the future.

psychedelicious commented:
> created `actionContextMiddleware`, which is responsible for injecting context into the action. Context is extracted by canvas and tab slices and used for action routing

For best performance, it's best to minimize the number of middlewares. Each middleware is run for every redux action and adds level(s) to the call stack for each action. The redux team suggests adding a listener to the listener middleware instead of adding a new middleware as a best practice. It's probably not problematic for Invoke but is something to keep in mind.

Can definitely leave UX/visual stuff for a pre-release. I will review the latest changes to the PR later today.

blessedcoolant commented:
Is it possible to sync up this PR with main because stuff has changed on that end and there's a ton of conflicts that prevents the app from booting completely to test. I don't want to resolve those conflicts myself coz I haven't gone through your changes yet. It'll help me start looking at the PR. Thank you.

Commit:
multi-canvas Redux refactor

Commit:
selectAllQueueItems selector memoization fixed

csehatt741 commented:
@blessedcoolant, I've rebased this branch onto `main`

csehatt741 commented:
@lstein, I believe that performance issue has also been fixed

blessedcoolant commented:
Started going through the PR. It's a massive one. Great job overall. Here's my initial thoughts on it all as I test it out. I might edit the post later to add as I keep diving into it further.

- [Bug] Creating the new canvas doesn't initialize the canvas to a functional state. I have to manually select a model to trigger the LoRA detection and other settings. This needs to be fixed.
- [Functionality] Right now, creating a new canvas creates a completely new entity in regards to everything -- right from prompt to settings to canvas layers. I feel this needs to be another feature entirely. The default "**New Canvas**" should carry over all settings and only the layer structure should be fresh and new. This is how most art applications do it and I think having that consistency would be good. If we want to provide the user with an option to completely have independent settings (which is the current functionality) .. let is provide it as another option and call it as "**Create Workspace**" or something along those lines. This would probably also fix the noticeable delay in shifting from one canvas to another coz the entire UI does not have to re-render again. But we'll get to that later.
- [UI] The new canvas I feel should be at the end of the canvas title bar and not the start.
- [UI] The edit title of the canvas can be moved next to the close coz it leads to many accidental clicks on it when selecting tabs. We could even leave it out coz double click does it but it might be visual feedback on the interface to have it. But moving it to the place where the close canvas icon exists seems like a better move to me.
-  When I create too many canvases and add long names, this happens. I think we truncate the title length and then instead of just constantly making the title tabs smaller and smaller, add a navigation button to go through open tabs instead.
<img width="783" height="107" alt="opera_PExIskPpPs" src="https://github.com/user-attachments/assets/e303df71-cdc0-4f71-ba2e-3d518c0847cd" />
If I add 10 more here, clicking on them will become super hard.
<img width="1610" height="96" alt="opera_5L8oTEBRts" src="https://github.com/user-attachments/assets/2200aab5-b7b7-46e5-96ae-241a5afc8ace" />

- [ Bug / feature?] Canvases I create are not persistent when I refresh the tab? Intended or still WIP?

I'll add more thing as I test. The number of code base change is massive. I'll trust the type checks for the time being. We'll do a full comb through once when all the functionality is in place. I think that is the better way to approach this given I have been out of touch with all the changes made over the past year.

csehatt741 commented:
Though, many files are involved in this PR, mainly due to selector refactorings, but the most important changes are in `store.ts` and canvas related slices, which should be reviewed thoroughly due to their impact on performance.
- I haven't fully finished this PR and there are some known minor regressions
- the feature has already been discussed to completely isolate canvas instances from each other: https://discord.com/channels/1020123559063990373/1149513625321603162/1426797542301241364
- the UX has also been decided to be finalized during prerelease and the current implementation is only for testing: https://github.com/invoke-ai/InvokeAI/pull/8553#issuecomment-3369528211
- persistence should not be an issue. Are there any errors in the console?
